packages/shader-lab/src/Error.ts

```diff
@@ -1,19 +1,79 @@
 // #if _EDITOR
-import { ShaderRange } from "./common";
+import { Logger } from "@galacean/engine";
+import { ShaderPosition, ShaderRange } from "./common";
 
 export abstract class GSError extends Error {
-  readonly loc: ShaderRange;
+  static wrappingLineCount = 2;
 
-  constructor(message: string, loc: ShaderRange, cause?: Error) {
+  readonly loc: ShaderRange | ShaderPosition;
+  readonly source: string;
+  readonly file?: string;
+
+  constructor(message: string, loc: ShaderRange | ShaderPosition, source: string, file?: string, cause?: Error) {
     super(message, { cause });
     this.loc = loc;
+    this.source = source;
+    this.file = file;
+  }
+
+  log(_source?: string): void {
+    if (!Logger.enable) return;
+    let start: ShaderPosition, end: ShaderPosition;
+    const { message, loc, source: originSource } = this;
+    let source = originSource;
+    if (_source) source = _source;
+    if (!source) {
+      Logger.error(message);
+      return;
+    }
+
+    if (loc instanceof ShaderPosition) {
+      start = end = loc;
+    } else {
+      start = loc.start;
+      end = loc.end;
+    }
+    const lines = source.split("\n");
+
+    let diagnosticMessage = `${this.name}: ${message}\n\n`;
+    const lineSplit = "|···";
+
+    for (let i = 0; i <= end.line + GSError.wrappingLineCount; i++) {
+      if (i < start.line - GSError.wrappingLineCount) continue;
+
+      diagnosticMessage += lineSplit;
+      diagnosticMessage += `${lines[i]}\n`;
+      const paddingLength = lineSplit.length + start.column;
+      const remarkLength = Math.max(end.column - start.column, 1);
+      if (i >= start.line && i <= end.line) {
+        diagnosticMessage += " ".repeat(paddingLength);
+        diagnosticMessage += "^".repeat(remarkLength);
+        diagnosticMessage += "\n";
+      }
+    }
+
+    Logger.error(diagnosticMessage);
+  }
+}
+
+export class PreprocessorError extends GSError {
+  constructor(message: string, loc: ShaderRange | ShaderPosition, source: string, file?: string, cause?: Error) {
+    super(message, loc, source, file, cause);
+    this.name = "PreprocessorError";
+  }
+}
```

Comment:
Tell me what is the value of using abstract classes？

Reply:
done